modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsWriter.java

```diff
@@ -0,0 +1,580 @@
+
+        /** {@code True} if the small buffer warning message logged. */
+        private final AtomicBoolean smallBufLogged = new AtomicBoolean();
+
+        /** {@code True} if worker stopped due to maximum file size reached. */
+        private final AtomicBoolean stopByMaxSize = new AtomicBoolean();
+
+        /**
+         * @param ctx Kernal context.
+         * @param fileIo Performance statistics file I/O.
+         * @param maxFileSize Maximum file size in bytes.
+         * @param bufferSize Off heap buffer size in bytes.
+         * @param flushBatchSize Minimal batch size to flush in bytes.
+         * @param log Logger.
+         */
+        FileWriter(GridKernalContext ctx, FileIO fileIo, long maxFileSize, int bufferSize, int flushBatchSize,
+            IgniteLogger log) {
+            super(ctx.igniteInstanceName(), "performance-statistics-writer%" + ctx.igniteInstanceName(), log);
+
+            this.fileIo = fileIo;
+            this.flushBatchSize = flushBatchSize;
+
+            ringByteBuffer = new SegmentedRingByteBuffer(bufferSize, maxFileSize, BufferMode.DIRECT);
+
+            ringByteBuffer.init(0);
+        }
+
+        /** {@inheritDoc} */
+        @Override protected void body() throws InterruptedException, IgniteInterruptedCheckedException {
+            while (!isCancelled()) {
+                blockingSectionBegin();
+
+                try {
+                    synchronized (this) {
+                        while (readyForFlushSize.get() < flushBatchSize && !isCancelled())
+                            wait();
+                    }
+                }
+                finally {
+                    blockingSectionEnd();
+                }
+
+                flushBuffer();
+            }
+
+            fileWriter = null;
+
+            ringByteBuffer.close();
+
+            // Make sure that all producers released their buffers to safe deallocate memory.
+            flushBuffer();
+
+            ringByteBuffer.free();
+
+            U.closeQuiet(fileIo);
+
+            cachedStrings.clear();
+
+            log.info("Performance statistics writer stopped.");
+        }
+
+        /** @return {@code True} if string hash code is cached. {@code False} if need write string.  */
+        boolean stringCached(String str) {
+            boolean cached = cachedStrings.contains(str.hashCode());
+
+            if (!cached)
+                cachedStrings.add(str.hashCode());
+
+            return cached;
+        }
+
+        /** @return Write segment.*/
+        SegmentedRingByteBuffer.WriteSegment writeSegment(int size) {
+            SegmentedRingByteBuffer.WriteSegment seg = ringByteBuffer.offer(size);
+
+            if (seg != null) {
+                int readySize = readyForFlushSize.addAndGet(size);
+
+                if (readySize >= DFLT_FLUSH_SIZE) {
+                    synchronized (this) {
+                        notify();
```

Comment:
Can you please, add a comment to clarify that this `notify` required to start writing data to the file.

Reply:
Done.